File: mvvmbase/src/main/java/com/bing/mvvmbase/base/recycleview/BaseRecycleViewFragment.java
package com.bing.mvvmbase.base.recycleview;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.bing.mvvmbase.R;
import com.bing.mvvmbase.base.BaseViewModel;
import com.bing.mvvmbase.model.datawrapper.Status;
import com.bing.mvvmbase.utils.DynamicTimeFormat;
import com.bing.statuslayout.StatusLayout;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.header.ClassicsHeader;
import com.scwang.smartrefresh.layout.listener.OnRefreshListener;

import java.util.Date;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.RecyclerView;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public abstract class BaseRecycleViewFragment<DB extends ViewDataBinding, VM extends BaseViewModel, AVM extends BaseViewModel, AD extends BaseRecycleViewAdapter, T> extends Fragment implements View.OnClickListener {
	protected DB mBinding;
	protected VM mViewModel;
	protected AVM mActivityViewModel;
	protected CompositeDisposable mCompositeDisposable = new CompositeDisposable();

	protected RecyclerView mRecyclerView;
	protected RefreshLayout mRefreshLayout;
	protected ClassicsHeader mClassicsHeader;
	protected AD mAdapter;
	protected StatusLayout mStatusLayout;

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		if (getArguments() != null) {
			handleArguments();
		}
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		mBinding = DataBindingUtil.inflate(inflater, layoutId(), container, false);
		initView();
		return mBinding.getRoot();
	}

	@Override
	public void onActivityCreated(@Nullable Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);
		initActivityViewModel();
		initViewModel();
		getLifecycle().addObserver(mViewModel);
		bindAndObserve();
	}

	/**
	 * 此时还没有viewModel
	 */
	protected void initView() {
		initStatusLayout();
		initRefreshLayout();
		initRefreshHeader();
		initRecycleView();
	}

	protected abstract void initStatusLayout();
	protected void initRefreshLayout() {
		mRefreshLayout = getRefreshLayout();
		mRefreshLayout.setEnableLoadMore(false);
		mRefreshLayout.setEnableOverScrollDrag(true);//是否启用越界拖动
		mRefreshLayout.setOnRefreshListener(new OnRefreshListener() {
			@Override
			public void onRefresh(@NonNull RefreshLayout refreshLayout) {
				refresh(refreshLayout);
			}
		});
	}

	protected abstract SmartRefreshLayout getRefreshLayout();
	protected abstract void refresh(@NonNull RefreshLayout refreshLayout);

	protected void initRefreshHeader() {
		mClassicsHeader = (ClassicsHeader) mRefreshLayout.getRefreshHeader();
		mClassicsHeader.setTimeFormat(new DynamicTimeFormat(getString(R.string.refresh_at) + " %s"));
	}

	protected void initRecycleView() {
		mRecyclerView = getRecyclerView();
		mRecyclerView.setLayoutManager(getLayoutManager());
		mRecyclerView.setItemAnimator(getItemAnimator());
		initAdapter();
		mRecyclerView.setAdapter(mAdapter);
	}

	protected abstract RecyclerView getRecyclerView();
	protected abstract RecyclerView.LayoutManager getLayoutManager();

	protected RecyclerView.ItemAnimator getItemAnimator() {
		return new DefaultItemAnimator();
	}
	protected abstract void initAdapter();
	protected abstract MutableLiveData<Status> getNetworkState();
	protected abstract MutableLiveData<Status> getRefreshState();
	protected abstract LiveData<List<T>> getData();

	protected abstract void handleArguments();
	protected abstract int layoutId();

	protected abstract void initActivityViewModel();
	protected abstract void initViewModel();

	protected void bindAndObserve() {
		getNetworkState().observe(this, new Observer<Status>() {
			@Override
			public void onChanged(@Nullable Status status) {
				if (status == null) {
					return;
				}
				switch (status) {
					case LOADING:
						mStatusLayout.showViewByStatus(StatusLayout.STATUS_LOADING);
						break;

					case SUCCESS:
						mStatusLayout.showViewByStatus(StatusLayout.STATUS_CONTENT);
						break;

					case ERROR:
						mStatusLayout.showViewByStatus(StatusLayout.STATUS_ERROR);
						break;

					case EMPTY:
						mStatusLayout.showViewByStatus(StatusLayout.STATUS_EMPTY);
						break;

					case NO_NETWORK:
						mStatusLayout.showViewByStatus(StatusLayout.STATUS_NO_NETWORK);
						break;

					default:
						break;
				}
			}
		});
		getData().observe(this, new Observer<List<T>>() {
			@Override
			public void onChanged(List<T> list) {
				mAdapter.setData(list);
			}
		});
		getRefreshState().observe(this, new Observer<Status>() {
			@Override
			public void onChanged(Status status) {
				if (status != Status.LOADING) {
					mRefreshLayout.finishRefresh(status == Status.SUCCESS);
					mClassicsHeader.setLastUpdateTime(new Date(System.currentTimeMillis()));
				}
			}
		});
	}

	protected void addOnClickListener(@NonNull View... views) {
		for (View view : views) {
			view.setOnClickListener(this);
		}
	}

	@Override
	public void onClick(View v) {

	}

	protected void addDisposable(Disposable disposable) {
		mCompositeDisposable.add(disposable);
	}

	@Override
	public void onDestroy() {
		mCompositeDisposable.dispose();
		super.onDestroy();
	}
}

File: mvvmbase/src/main/java/com/bing/mvvmbase/base/recycleview/BaseRecycleViewFragment.kt
package com.bing.mvvmbase.base.recycleview

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.bing.mvvmbase.R
import com.bing.mvvmbase.base.BaseViewModel
import com.bing.mvvmbase.base.IsSame
import com.bing.mvvmbase.model.datawrapper.Status
import com.bing.mvvmbase.utils.DynamicTimeFormat
import com.bing.statuslayout.StatusLayout
import com.scwang.smartrefresh.layout.SmartRefreshLayout
import com.scwang.smartrefresh.layout.api.RefreshLayout
import com.scwang.smartrefresh.layout.header.ClassicsHeader

import java.util.Date
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.fragment.app.Fragment
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.RecyclerView
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

abstract class BaseRecycleViewFragment<DB : ViewDataBinding, VM : BaseViewModel, AVM : BaseViewModel,  AD : BaseRecycleViewAdapter<T, *>, T : IsSame> : Fragment(), View.OnClickListener {
        lateinit var mBinding: DB
        lateinit var mViewModel: VM
        lateinit var mActivityViewModel: AVM
        protected var mCompositeDisposable = CompositeDisposable()

        lateinit var mRecyclerView: RecyclerView
        lateinit var mRefreshLayout: RefreshLayout
        lateinit var mClassicsHeader: ClassicsHeader
        lateinit var mAdapter: AD
        lateinit var mStatusLayout: StatusLayout

        protected abstract val refreshLayout: SmartRefreshLayout

        protected abstract val recyclerView: RecyclerView
        protected abstract val layoutManager: RecyclerView.LayoutManager

        protected var itemAnimator = DefaultItemAnimator()
        protected abstract val networkState: MutableLiveData<Status>
        protected abstract val refreshState: MutableLiveData<Status>
        protected abstract val data: LiveData<List<T>>

        override fun onCreate(savedInstanceState: Bundle?) {
                super.onCreate(savedInstanceState)
                if (arguments != null) {
                        handleArguments()
                }
        }

        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
                mBinding = DataBindingUtil.inflate(inflater, layoutId(), container, false)
                initView()
                return mBinding.root
        }

        override fun onActivityCreated(savedInstanceState: Bundle?) {
                super.onActivityCreated(savedInstanceState)
                mActivityViewModel = ViewModelProviders.of(activity!!).get(mActivityViewModel.javaClass)
                mViewModel = ViewModelProviders.of(this).get(mViewModel.javaClass)
                lifecycle.addObserver(mViewModel)
                bindAndObserve()
        }

        /**
         * 此时还没有viewModel
         */
        protected fun initView() {
                initStatusLayout()
                initRefreshLayout()
                initRefreshHeader()
                initRecycleView()
        }

        protected abstract fun initStatusLayout()
        protected fun initRefreshLayout() {
                mRefreshLayout = refreshLayout
                mRefreshLayout.setEnableLoadMore(false)
                mRefreshLayout.setEnableOverScrollDrag(true)//是否启用越界拖动
                mRefreshLayout.setOnRefreshListener { refreshLayout -> refresh(refreshLayout) }
        }

        protected abstract fun refresh(refreshLayout: RefreshLayout)

        protected fun initRefreshHeader() {
                mClassicsHeader = mRefreshLayout.refreshHeader as ClassicsHeader
                mClassicsHeader.setTimeFormat(DynamicTimeFormat(getString(R.string.refresh_at) + " %s"))
        }

        protected fun initRecycleView() {
                mRecyclerView = recyclerView
                mRecyclerView.layoutManager = layoutManager
                mRecyclerView.itemAnimator = itemAnimator
                initAdapter()
                mRecyclerView.adapter = mAdapter
        }

        protected abstract fun initAdapter()

        protected abstract fun handleArguments()
        protected abstract fun layoutId(): Int

        protected fun bindAndObserve() {
                networkState.observe(this, Observer { status ->
                        if (status == null) {
                                return@Observer
                        }
                        when (status) {
                                Status.LOADING -> mStatusLayout.showViewByStatus(StatusLayout.STATUS_LOADING)

                                Status.SUCCESS -> mStatusLayout.showViewByStatus(StatusLayout.STATUS_CONTENT)

                                Status.ERROR -> mStatusLayout.showViewByStatus(StatusLayout.STATUS_ERROR)

                                Status.EMPTY -> mStatusLayout.showViewByStatus(StatusLayout.STATUS_EMPTY)

                                Status.NO_NETWORK -> mStatusLayout.showViewByStatus(StatusLayout.STATUS_NO_NETWORK)

                                else -> {
                                }
                        }
                })
                data.observe(this, Observer { list -> mAdapter.data = list })
                refreshState.observe(this, Observer { status ->
                        if (status != Status.LOADING) {
                                mRefreshLayout.finishRefresh(status == Status.SUCCESS)
                                mClassicsHeader.setLastUpdateTime(Date(System.currentTimeMillis()))
                        }
                })
        }

        protected fun addOnClickListener(vararg views: View) {
                for (view in views) {
                        view.setOnClickListener(this)
                }
        }

        override fun onClick(v: View) {

        }

        protected fun addDisposable(disposable: Disposable) {
                mCompositeDisposable.add(disposable)
        }

        override fun onDestroy() {
                mCompositeDisposable.dispose()
                super.onDestroy()
        }
}
